src/SourceBuild/content/eng/Publishing.props

```diff
@@ -0,0 +1,92 @@
+    and shouldn't be pushed to build storage.
+  -->
+  <ItemGroup>
+    <ArtifactVisibilityToPublish Include="Internal;External" />
+  </ItemGroup>
+
+  <!-- Build task assembly paths -->
+  <PropertyGroup>
+    <MicrosoftDotNetUnifiedBuildTasksAssembly>$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'Microsoft.DotNet.UnifiedBuild.Tasks', '$(Configuration)', 'Microsoft.DotNet.UnifiedBuild.Tasks.dll'))</MicrosoftDotNetUnifiedBuildTasksAssembly>
+  </PropertyGroup>
+
+  <UsingTask TaskName="Microsoft.DotNet.UnifiedBuild.Tasks.GetKnownArtifactsFromAssetManifests" AssemblyFile="$(MicrosoftDotNetUnifiedBuildTasksAssembly)" TaskFactory="TaskHostFactory" />
+
+  <Target Name="DiscoverArtifacts">
+    <ItemGroup>
+      <!-- Repo manifests from individual repos -->
+      <RepoManifests Include="$(AssetManifestsIntermediateDir)/**/*.xml" />
+    </ItemGroup>
+    <!-- Get produced packages and assets from the manifests -->
+    <GetKnownArtifactsFromAssetManifests AssetManifests="@(RepoManifests)">
+      <Output TaskParameter="KnownPackages" ItemName="ProducedPackage" />
+      <Output TaskParameter="KnownBlobs" ItemName="ProducedAsset" />
+    </GetKnownArtifactsFromAssetManifests>
+    <ItemGroup>
+      <ProducedPackage>
+        <IsShipping Condition="'%(ProducedPackage.NonShipping)' != 'true'">true</IsShipping>
+        <IsShipping Condition="'%(ProducedPackage.NonShipping)' == 'true'">false</IsShipping>
+        <Kind>Package</Kind>
+        <PublishFlatContainer>false</PublishFlatContainer>
+      </ProducedPackage>
+      <ProducedPackage>
+        <ShippingFolder Condition="'%(ProducedPackage.NonShipping)' != 'true'">Shipping</ShippingFolder>
+        <ShippingFolder Condition="'%(ProducedPackage.NonShipping)' == 'true'">NonShipping</ShippingFolder>
+      </ProducedPackage>
+
+      <ReferencePackage Include="@(ProducedPackage->'$(ReferencePackagesDir)%(Identity).%(Version).nupkg')" Condition="'%(RepoOrigin)' == 'source-build-reference-packages'" />
+    </ItemGroup>
+
+    <!-- Create Artifact items for produced packages and assets -->
+    <ItemGroup>
+      <Artifact Include="@(ProducedPackage->'$(ArtifactsPackagesDir)%(ShippingFolder)/%(RepoOrigin)/%(Identity).%(Version).nupkg')" Condition="'%(RepoOrigin)' != 'source-build-reference-packages'" />
+      <Artifact Include="@(ReferencePackage)" RepoOrigin="SourceBuildReferencePackages" />
+      <Artifact Include="@(ProducedAsset->'$(ArtifactsAssetsDir)%(Identity)')">
+        <Kind>Asset</Kind>
+      </Artifact>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="FinishSourceBuild" AfterTargets="Publish" Condition="'$(DotNetBuildSourceOnly)' == 'true'">
+    <MSBuild
+      Projects="$(RepositoryEngineeringDir)finish-source-only.proj"
+      Targets="Restore"
+      Properties="MSBuildRestoreSessionId=$([System.Guid]::NewGuid())" />
+    <MSBuild
+      Projects="$(RepositoryEngineeringDir)finish-source-only.proj" />
+  </Target>
```

Comment:
Most if not all of `finish-source-only.proj` project's logic is unrelated to publishing. It scans for prebuilts and runs poison checks. It also creates archives that ares secondary outputs of the VMR build for source-build. I would keep this in the current `Build` target.

cc @NikolaMilosavljevic @MichaelSimons

Reply:
That's only the `CreatePrivateSourceBuiltArtifactsArchive` target and its dependencies. Everything else isn't tied to the manifest and should run as part of the build target.

I would pull that infra out and move it into Publishing.props (or a separate project) and leave everything else in finish-source-only.proj to avoid the weirdness around just-in-time restore and build.

Reply:
As far as I can tell, almost everything in `finish-source-only` depends on the published artifacts in some way or other.

- `CreateSdkSymbolsTarball` depedns on the final published .NET SDK tarball.
- `ReportPoisonUsage` depends on the list of published assets
- `CreatePrivateSourceBuiltArtifactsArchive` depends on the published assets and the merged asset manifest

Only the prebuilts support doesn't use it.

I'll pull it out to another .props file that's imported into in `Publishing.props`

Reply:
> CreateSdkSymbolsTarball depedns on the final published .NET SDK tarball.

Would you mind elaborating on that?. The build target already produces all the inputs to that target.

Reply:
CreateSdkSymbolsTarball takes in an expanded SDK tarball layout (which it pulls from the published artifacts) as well as the unified symbols layout (generated in finish-source-only). From those inputs, it selects the symbols that correspond to files in the SDK layout.

Reply:
Looks like the SDK tarball that is expanded comes from the pre-publish layout, so we can move the symbols logic to pre-publish.